code/obj/random_spawners.dm

```diff
@@ -30,24 +32,45 @@
 				if (isnum(guaranteed[new_item]))
 					amt = abs(guaranteed[new_item])
 				for (amt, amt>0, amt--)
-					new new_item(src.loc)
+					closet_check_spawn(new_item)
 
 		if (!islist(src.items2spawn) || !src.items2spawn.len)
 			logTheThing("debug", src, null, "has an invalid items2spawn list")
 			return
-
+		if (rare_chance)
+			if (!islist(src.rare_items2spawn) || !src.rare_items2spawn.len)
+				logTheThing("debug", src, null, "has an invalid rare_items2spawn list")
+				return
 		if (amt2spawn == 0)
 			amt2spawn = rand(min_amt2spawn, max_amt2spawn)
 		if (amt2spawn == 0) // If for whatever reason we still end up with 0...
 			return
 		for (amt2spawn, amt2spawn>0, amt2spawn--)
-			var/obj/new_item = pick(src.items2spawn)
+			// first, decide whether or not we will spawn a rare item!
+			var/list/item_list = list()
+			if (prob(rare_chance))
+				if (rare_items2spawn)
+					item_list = rare_items2spawn
+				else
+					logTheThing("debug", src, null, "has an invalid rare spawn list, [rare_items2spawn]")
+					DEBUG_MESSAGE("[src] has an invalid rare spawn list, [rare_items2spawn]")
+					continue
+			else
+				item_list = items2spawn
+			var/obj/new_item = pick(item_list)
 			if (!ispath(new_item))
 				logTheThing("debug", src, null, "has a non-path item in its spawn list, [new_item]")
 				DEBUG_MESSAGE("[src] has a non-path item in its spawn list, [new_item]")
 				continue
-			new new_item(src.loc)
 
+			closet_check_spawn(new_item)
+
+	proc/closet_check_spawn(var/obj/item/new_item)
+		var/obj/storage/S = locate(/obj/storage) in src.loc.contents
```

Comment:
```suggestion
		var/obj/storage/S = locate(/obj/storage) in src.loc
```
`contents` is implicit in these cases.

Reply:
Oh! @AwkwardDryad made this bit of code too.